extensions/amp-story/1.0/background-blur.js

```diff
@@ -86,58 +89,130 @@ export class BackgroundBlur {
-          const innerVideoEl = mediaEl.querySelector('video');
-          const alreadyHasData = innerVideoEl.readyState >= HAVE_CURRENT_DATA;
-          if (alreadyHasData) {
-            this.animate_(innerVideoEl);
-            return;
-          }
-          // If video doesnt have data, render from the poster image.
-          const posterSrc = mediaEl.getAttribute('poster');
-          if (!posterSrc) {
-            this.animate_();
-            user().info(CLASS_NAME, 'No "poster" attribute on amp-video.');
-            return;
-          }
-          const img = new Image();
-          img.onload = () => this.animate_(img);
-          img.src = posterSrc;
-        },
-        () => {
-          user().error(CLASS_NAME, 'Failed to load the amp-img or amp-video.');
+    this.ensureMediaLoaded_(mediaEl).then((loadedMediaEl) => {
+      // If image:
+      if (loadedMediaEl.tagName === 'AMP-IMG') {
+        // First check if it has transparent pixels.
+        if (this.isTransparentGifOrPng_(loadedMediaEl)) {
+          // If transparent, try again with the next element in the array.
+          this.findAndBlurMediaEl_(this.getNextMediaEl_(loadedMediaEl));
+          return;
         }
-      );
+        this.animateBlur_(loadedMediaEl.querySelector('img'));
+        return;
+      }
+
+      // If video, render first frame or poster image.
+      const innerVideoEl = loadedMediaEl.querySelector('video');
+      const alreadyHasData = innerVideoEl.readyState >= HAVE_CURRENT_DATA;
+      if (alreadyHasData) {
+        this.animateBlur_(innerVideoEl);
+        return;
+      }
+      // If video doesnt have data, render from the poster image.
+      const posterSrc = loadedMediaEl.getAttribute('poster');
+      if (!posterSrc) {
+        this.animateBlur_();
+        user().info(CLASS_NAME, 'No "poster" attribute on amp-video.');
+        return;
+      }
+      const img = new Image();
+      img.onload = () => this.animateBlur_(img);
+      img.src = posterSrc;
+    });
+  }
+
+  /**
+   * @private
+   * @param {?Element} mediaEl
+   * @return {boolean}
+   */
+  isTransparentGifOrPng_(mediaEl) {
+    if (!this.isGifOrPng_(mediaEl)) {
+      return false;
+    }
+    const imgEl = mediaEl.querySelector('img');
+    const canvas = this.win_.document.createElement('canvas');
+    canvas.width = canvas.height = CANVAS_SIZE;
+    const context = canvas.getContext('2d');
+    context.drawImage(imgEl, 0, 0, CANVAS_SIZE, CANVAS_SIZE);
+    const imgData = context.getImageData(0, 0, CANVAS_SIZE, CANVAS_SIZE).data;
+    // Image data pixel values are in sets of 4: r, g, b, a.
+    // For this reason we increment in 4.
+    for (let i = 0; i < imgData.length; i += 4) {
+      const pixelAlphaVal = imgData[i + 3];
+      if (pixelAlphaVal < 255) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * @private
+   * @param {?Element} mediaEl
+   * @return {boolean}
+   */
+  isGifOrPng_(mediaEl) {
```

Comment:
This might be a bit weak, happy to test it and revisit if we notice too many issues. We can always try to get the mime type of the image but it's more expensive, so this is a good place to start.

Reply:
Agreed. It assumes the filename extension wasn't manually changed from `png` to `jpg`.
I feel like that would only happen if someone was intentionally abusing the function, and if they make something cool from it I want to see it!